ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/jsonrpc/JsonRpcSubscriptionManager.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.ide.jsonrpc;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import org.eclipse.che.api.core.jsonrpc.commons.RequestHandlerConfigurator;
+import org.eclipse.che.api.core.jsonrpc.commons.RequestTransmitter;
+import org.eclipse.jetty.util.ConcurrentHashSet;
+import org.slf4j.Logger;
+
+/**
+ * Client side class simplifies subscription management by providing a more convenient API for
+ * initiating subscription and reacting on events that comes from server side.
+ */
+public abstract class JsonRpcSubscriptionManager {
+  private static final Logger LOGGER = getLogger(JsonRpcSubscriptionManager.class);
+
+  private final RequestTransmitter requestTransmitter;
+  private final RequestHandlerConfigurator requestHandlerConfigurator;
+
+  private final Map<String, Set<Consumer>> methodsToConsumers = new ConcurrentHashMap<>();
```

Comment:
ConcurrentHashMap in GWT client? Maybe simple HashMap will be enough?

Reply:
Taking into account the way GWT works right now you're absolutely right, it is effectively enough to use simple hash map. However my preference here is to not rely on low level implementation features of GWT, for example if they change it in future releases and the behavior of hash and concurrent hash maps will be different we will need to rewrite the code. What do you think?

Reply:
As of what I see [here](http://www.gwtproject.org/doc/latest/RefJreEmulation.html#Package_java_util_concurrent) it should, so no worries.